crates/common/src/local_db/query/fetch_vaults/mod.rs

```diff
@@ -7,67 +10,70 @@ pub struct FetchVaultsArgs {
-pub fn build_fetch_vaults_query(chain_id: u32, args: &FetchVaultsArgs) -> String {
-    let sanitize_literal = |value: &str| value.replace('\'', "''");
-
-    let owner_values: Vec<String> = args
-        .owners
-        .iter()
-        .filter_map(|owner| {
-            let trimmed = owner.trim();
-            if trimmed.is_empty() {
-                None
-            } else {
-                let lowered = trimmed.to_lowercase();
-                Some(format!("'{}'", sanitize_literal(&lowered)))
-            }
-        })
-        .collect();
-    let filter_owners = if owner_values.is_empty() {
-        String::new()
-    } else {
-        format!("\nAND lower(o.owner) IN ({})\n", owner_values.join(", "))
-    };
-
-    let token_values: Vec<String> = args
-        .tokens
-        .iter()
-        .filter_map(|token| {
-            let trimmed = token.trim();
-            if trimmed.is_empty() {
-                None
-            } else {
-                let lowered = trimmed.to_lowercase();
-                Some(format!("'{}'", sanitize_literal(&lowered)))
-            }
-        })
-        .collect();
-    let filter_tokens = if token_values.is_empty() {
-        String::new()
-    } else {
-        format!("\nAND lower(o.token) IN ({})\n", token_values.join(", "))
-    };
-
-    const BALANCE_EXPR: &str = r"#COALESCE((
-    SELECT FLOAT_SUM(vd.delta)
-    FROM vault_deltas vd
-    WHERE vd.owner    = o.owner
-      AND vd.token    = o.token
-      AND vd.vault_id = o.vault_id
-  ), FLOAT_ZERO_HEX())
-        #";
-
-    let filter_hide_zero_balance = if args.hide_zero_balance {
-        format!("\nAND NOT FLOAT_IS_ZERO({expr})\n", expr = BALANCE_EXPR)
+const OWNERS_CLAUSE: &str = "/*OWNERS_CLAUSE*/";
+const OWNERS_CLAUSE_BODY: &str = "\nAND lower(o.owner) IN ({list})\n";
+
+const TOKENS_CLAUSE: &str = "/*TOKENS_CLAUSE*/";
+const TOKENS_CLAUSE_BODY: &str = "\nAND lower(o.token) IN ({list})\n";
+
+const HIDE_ZERO_BALANCE_CLAUSE: &str = "/*HIDE_ZERO_BALANCE*/";
+const HIDE_ZERO_BALANCE_BODY: &str =
+    "\nAND NOT FLOAT_IS_ZERO(\n    COALESCE((\n      SELECT FLOAT_SUM(vd.delta)\n      FROM vault_deltas vd\n      WHERE vd.owner    = o.owner\n        AND vd.token    = o.token\n        AND vd.vault_id = o.vault_id\n    ), FLOAT_ZERO_HEX())\n  )\n\n";
+
+pub fn build_fetch_vaults_stmt(
+    chain_id: u32,
+    args: &FetchVaultsArgs,
+) -> Result<SqlStatement, SqlBuildError> {
+    let mut stmt = SqlStatement::new(QUERY_TEMPLATE);
+    // ?1: chain id
+    stmt.push(SqlValue::I64(chain_id as i64));
+
+    // Owners list (trim, non-empty, lowercase) with order-preserving dedup
+    let mut owners: Vec<String> = Vec::new();
+    let mut seen: HashSet<String> = HashSet::new();
+    for owner in args.owners.iter() {
+        let t = owner.trim();
+        if t.is_empty() {
+            continue;
+        }
+        let lowered = t.to_ascii_lowercase();
+        if seen.insert(lowered.clone()) {
+            owners.push(lowered);
+        }
+    }
```

Comment:
nitpick: consider (requires itertools)

```suggestion
    let owners: Vec<String> = args.owners
        .iter()
        .map(|owner| owner.trim().to_ascii_lowercase())
        .filter(|s| !s.is_empty())
        .unique()
        .collect();
```